internal/endpoint/endpoint.go

```diff
@@ -2,59 +2,87 @@
 //
 // Copyright (C) 2018 IOTech Ltd
 // Copyright (c) 2019 Intel Corporation
+// Copyright (c) 2020 Dell Inc.
 //
 // SPDX-License-Identifier: Apache-2.0
 
 package endpoint
 
 import (
+	"context"
 	"fmt"
 	"os"
 	"sync"
 	"time"
 
-	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/interfaces"
 	"github.com/edgexfoundry/go-mod-registry/registry"
 )
 
 type Endpoint struct {
-	RegistryClient registry.Client
-	passFirstRun   bool
-	WG             *sync.WaitGroup
+	ctx            context.Context
+	wg             *sync.WaitGroup
+	RegistryClient *registry.Client
+	serviceKey     string // The key of the service as found in the service registry (e.g. Consul)
+	path           string // The path to the service's endpoint following port number in the URL
+	interval       int    // The interval in milliseconds governing how often the client polls to keep the endpoint current
+
 }
 
-func (e Endpoint) Monitor(params types.EndpointParams) chan string {
-	ch := make(chan string, 1)
-	ch <- e.fetch(params)
+func New(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	registryClient *registry.Client,
+	serviceKey string,
+	path string,
+	interval int) *Endpoint {
+
+	return &Endpoint{
+		ctx:            ctx,
+		wg:             wg,
+		RegistryClient: registryClient,
+		serviceKey:     serviceKey,
+		path:           path,
+		interval:       interval,
+	}
+}
+
+func (e Endpoint) Monitor() chan interfaces.URLStream {
+	ch := make(chan interfaces.URLStream, 1)
+	url, err := e.buildURL()
+	if err != nil {
+		_, _ = fmt.Fprintln(os.Stdout, err.Error())
+	}
+	ch <- interfaces.URLStream(url)
+
+	e.wg.Add(1)
 	go func() {
+		defer e.wg.Done()
 		for {
-			url, err := e.buildURL(params)
+			url, err := e.buildURL()
 			if err != nil {
-				fmt.Fprintln(os.Stdout, err.Error())
+				_, _ = fmt.Fprintln(os.Stdout, err.Error())
+			}
+			ch <- interfaces.URLStream(url)
+			time.Sleep(time.Millisecond * time.Duration(e.interval))
+
+			select {
+			case <-e.ctx.Done():
```

Comment:
I guess this would block this for loop, because this context is never cancelled.
https://golang.org/pkg/context/#Background

Reply:
Once https://github.com/edgexfoundry/device-sdk-go/pull/432 is merged in, the context should be automatically cancelled on shutdown. Maybe we should sequence this PR after that one?

Reply:
sounds good to me.  Let's hold this PR for until #432 is merged.  The contract update is not a rush request, and contract is keeping update anyway.

Reply:
@brandonforster we need this PR for https://github.com/edgexfoundry/device-sdk-go/issues/465
May we resolve this before #432?
Also, could you please upgrade contract to v0.1.52?
Thanks.

Reply:
I'm not sure where the ctx would be cancel in #432, @brandonforster could you elaborate more on this ?

Reply:
The context is intended to be canceled by the bootstrap system and allow for a graceful shutdown. This context is shared by all services using the bootstrap handlers, and when one handler sends `ctx.Done()`, everything should start attempting to gracefully shut down.

What we do not want are go routines running forever until the Go runtime forces them to quit.

I believe @michaelestrin can also add some context here as the primary bootstrap author. All my knowledge is second hand from observing him and his work.

Reply:
On target @brandonforster!